pkg/gadget-collection/gadgets/advise/seccomp/gadget.go

```diff
@@ -397,21 +406,36 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
 		// outside of the gadget control. Make a copy for the callback.
 		traceCopy := trace.DeepCopy()
 
-		// Subscribes to container termination events in order to
-		// generate a SeccompProfile when a container terminates. We
-		// don't need the list of existing containers, so the return
-		// value is ignored.
-		_ = t.helpers.Subscribe(
+		getContainerOwnerReference := func(c *containercollection.Container) {
+			_, err := c.GetOwnerReference()
+			if err != nil && !errors.Is(err, rest.ErrNotInCluster) {
+				log.Warnf("Failed to get owner reference of %s/%s/%s",
+					c.Namespace, c.Podname, c.Name)
+			}
+		}
+
+		// Subscribe to container creation and termination
+		// events. Termination is used to generate a
+		// SeccompProfile when a container terminates. Creation
+		// is used to fetch the owner reference of the
+		// containers to be sure this field is set when the
+		// container terminates.
+		containers := t.helpers.Subscribe(
 			genPubSubKey(trace.ObjectMeta.Namespace+"/"+trace.ObjectMeta.Name),
 			*gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter),
 			func(event containercollection.PubSubEvent) {
-				// Ignore container creation events.
-				if event.Type != containercollection.EventTypeRemoveContainer {
-					return
+				switch event.Type {
+				case containercollection.EventTypeAddContainer:
+					getContainerOwnerReference(event.Container)
```

Comment:
I tested it with 4 use-cases:

**Generate profile on-demand**: Working, but it is retrieving twice the owner references from the system.

1. Start gadget: 
```bash
$ ./kubectl-gadget advise seccomp-profile start -n default -p web-0 -m seccomp-profile
```
2. Deploy app: 
```bash
$ k apply -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
```
3. Stop gadget
4. Check that profile contains the owner ref annotations:
```bash
$ k get sp -n gadget web-0 -o jsonpath='{.metadata.annotations}' | jq
{
  ...
  "seccomp.gadget.kinvolk.io/ownerReference-APIVersion": "apps/v1",
  "seccomp.gadget.kinvolk.io/ownerReference-Kind": "StatefulSet",
  "seccomp.gadget.kinvolk.io/ownerReference-Name": "web",
  "seccomp.gadget.kinvolk.io/ownerReference-UID": "7aaaa2c1-13cd-44b5-a702-7b3402d8881b",
  ...
}
```
5. Clean up
```bash
$ k delete -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
$ k delete sp -n gadget web-0
```

**Generate profile upon pod termination**: No profiles are generated

 1. Start gadget: 
```bash
$ ./kubectl-gadget advise seccomp-profile start -n default -p web-0 -m seccomp-profile
```
2. Deploy app: 
```bash
$ k apply -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
```
3. Delete app:
```bash
$ k delete -f https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/application/web/web.yaml
```

If I repeat the two tests but exchanging steps 1 and 2 (deploying the app before starting the gadget), the test works correctly, and the owner reference is retrieved only once.

IMO, the enrichment of the owner reference is not saved when operating on the `event.Container` but only when done on the container returned in line 423. It looks like the event.Container is not actually the correct pointer. I took a look, and I think it is because [here](https://github.com/kinvolk/inspektor-gadget/blob/main/pkg/container-collection/container-collection.go#L148) we are making a copy. WDYT?

Reply:
Good catch. We should add some tests to be sure this pusub mechanism is not passing a copy of the container (in another PR).

Reply:
We should also add these 4 use cases to the integration tests.

Reply:
The four use cases are now working properly.